### PR TITLE
Determining possibly modified registers statically

### PR DESCRIPTION
**Pull request by shigoel**
### Description:

Rudimentary implementation of statically determining the GPR/SFPs that may be modified by the program.

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine?

Yes

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Arm/Cfg/Cfg.lean [hunk 1]
```diff
@@ -9,13 +9,17 @@ namespace Cfg
 
 open BitVec
 
-/-- Conditions under which a branch is taken; this is a function that
-takes a state as input, and returns a boolean. -/
+/--
+Conditions under which a branch is taken; this is a function that
+takes a state as input, and returns a boolean.
```
Comment (bollu) [inline Arm/Cfg/Cfg.lean:14]:
```suggestion
takes a state as input, and returns whether the branch was taken.
```

File: Arm/Cfg/Cfg.lean [hunk 2]
```diff
@@ -120,9 +132,9 @@ private def loop_detected (from_inst : InstType) (to_insts : List InstType) : IO
       have h' : next.length > 0 := by simp_all
       pure (some { guard := from_inst, target := to_inst, next := next[0]'h' })
     else
-      throw (IO.userError
-       ("We expected exactly one Seq instruction in the control-flow graph for this entry. " ++
-        "Instead, we found {next.length}."))
+      .error
```
Comment (bollu) [inline Arm/Cfg/Cfg.lean:135]: Nice, it's a pure function now!

File: Arm/Cfg/Cfg.lean [hunk 3]
```diff
@@ -88,25 +94,31 @@ deriving Repr
...
 
 In that case, `to_pc` has the loop target instruction and `from_pc`
 has the loop guard instruction. The first instruction that will be
 executed after the loop terminates will also be a member of `to_insts`
-of type `Seq`. -/
-private def loop_detected (from_inst : InstType) (to_insts : List InstType) : IO (Option LoopInfo) :=
+of type `Seq`.
+-/
+private def loop_detected (from_inst : InstType) (to_insts : List InstType) :
+  Except String (Option LoopInfo) :=
```
Comment (bollu) [inline Arm/Cfg/Cfg.lean:121]:
```suggestion
  Except Format (Option LoopInfo) :=
```

If we use a `Format`, we can write richer, structured logs with combinators like `nest, group`. We can then call `toString : Format -> String` on a `Format` to make a string out of it.

File: Arm/Cfg/Cfg.lean [hunk 4]
```diff
@@ -141,20 +153,38 @@ private def addEntry (from_inst : InstType) (to_insts : List InstType)
...
---         arm_inst -- current Arm instruction
---         cfg -- the control-flow graph
--- outputs: haltp : Bool -- whether the program halts
---          cfg : Cfg -- the updated control-flow graph
-protected def addArmInstToCfg (pc : BitVec 64) (arm_inst : ArmInst) (cfg : Cfg)
-   : IO (Bool × Cfg) := do
+  .error
+    s!"[ForwardGraph] Implementation Error: graph already contains \
+    an entry with PC {InstType.pc from_inst}! \
+    Here is the graph: ${cfg.graph}."
```
Comment (bollu) [inline Arm/Cfg/Cfg.lean:172]:
```suggestion
    f!"[ForwardGraph] Implementation Error: graph already contains \
    an entry with PC {InstType.pc from_inst}! \
    Here is the graph: ${Format.indentD cfg.graph}."
```

**Review by bollu — COMMENTED** [review]
Nice, I like how this is shaping up!

File: Tests/SHA2/SHA512ProgramTest.lean [hunk 5]
```diff
@@ -13,15 +13,27 @@ section SHA512ProgramTest
...
-info: #[(0,
+info: #[RegType.GPR 0x01#5, RegType.GPR 0x02#5, RegType.GPR 0x03#5, RegType.GPR 0x04#5, RegType.SFP 0x00#5,
+  RegType.SFP 0x01#5, RegType.SFP 0x02#5, RegType.SFP 0x03#5, RegType.SFP 0x04#5, RegType.SFP 0x05#5,
+  RegType.SFP 0x06#5, RegType.SFP 0x07#5, RegType.SFP 0x10#5, RegType.SFP 0x11#5, RegType.SFP 0x12#5,
+  RegType.SFP 0x13#5, RegType.SFP 0x14#5, RegType.SFP 0x15#5, RegType.SFP 0x16#5, RegType.SFP 0x17#5,
+  RegType.SFP 0x18#5, RegType.SFP 0x19#5, RegType.SFP 0x1a#5, RegType.SFP 0x1b#5, RegType.SFP 0x1c#5,
+  RegType.SFP 0x1d#5]
+-/
+#guard_msgs in
+#eval ((Cfg.create' 0x126500#64 0x126c90#64 SHA512.program).toOption).get!.maybe_modified_regs
```
Comment (shigoel) [inline Tests/SHA2/SHA512ProgramTest.lean:27]: @bollu Mind helping me simplify this after I embraced `Format`? I'd like this to look something like the following:

```
do let cfg ← Cfg.create SHA512.program; pure cfg.maybe_modified_regs
```

File: Proofs/SHA512/SHA512Loop.lean [hunk 6]
```diff
@@ -14,110 +16,288 @@ is satisfied.
...
-  -- assumption
-  -- done
-  sorry
+  simp only [←Memory.mem_eq_iff_read_mem_bytes_eq] at *
+  simp only [memory_rules] at *
+  simp at h_si_separate
+  simp only [h_si_ktbl, h_si_separate, minimal_theory]
+  have h_si_input_1 : (Memory.read_bytes 16 InputBase sprev.mem) =
+                      (Memory.read_bytes 16 InputBase si.mem) := by
+    -- simp_mem (config := {useOmegaToClose := true})
```
Comment (shigoel) [inline Proofs/SHA512/SHA512Loop.lean:181]: @bollu What am I doing wrong that `simp_mem` isn't helping here?

Comment (shigoel) [reply Proofs/SHA512/SHA512Loop.lean:209]: From a conversation with @bollu: looks like `omega` is generating a gigantic proof, which causes `simp_mem` to fail here. We'll table this for now.

**Pull request merged**